### PR TITLE
Fix #783 - 'respawn' event

### LINKED ISSUE #783: 'spawn' never gets triggered but 'update_health' event is received
it's possible your server doesn't send the update_health packet and hence spawn doesn't get emitted https://github.com/PrismarineJS/mineflayer/blob/7e3da452d6c75b44b612f8c174700699687d7495/lib/plugins/health.js#L11

_Originally posted by @rom1504 in https://github.com/PrismarineJS/mineflayer/issues/749#issuecomment-441462483_

---

This is my code:
```JavaScript
const mineflayer = require('mineflayer');
const navigatePlugin = require('mineflayer-navigate')(mineflayer);
const vec3 = require('vec3')

let bot = mineflayer.createBot({
    host: "localhost",
    port: 25565,
    username: "testing",
    version:"1.12.2"
});

bot.on('spawn', () => {
    bot.chat('I spawned, watch out!')
})

bot._client.on('update_health', (packet) => { 
    console.log(packet)
})
```

But even though I'm receiving the _update_health_ packet, the _spawn_ event never gets triggered.

### PR DESCRIPTION
**Pull request by ImHarvol**
Fix #783 by fixing 'respawn' event that didn't got triggered when the client connected to a server

### CODE REVIEW

File: lib/plugins/health.js [hunk 1]
```diff
@@ -8,6 +8,12 @@ function inject (bot) {
     bot.emit('respawn')
   })
 
+  bot._client.once('update_health', (packet) => {
+    if (bot.isAlive && packet.health > 0) {
```
Comment (wvffle) [inline lib/plugins/health.js]: `bot.isAlive` is always `true` as this event runs once on start

**Comment by ImHarvol** [comment]
> I think that you should add a test case for this issue

What do you mean by adding a test case?
Like in https://github.com/PrismarineJS/mineflayer/tree/master/test?

**Comment by wvffle** [comment]
Yes

File: test/internalTest.js [hunk 2]
```diff
@@ -60,6 +60,24 @@ mineflayer.supportedVersions.forEach((supportedVersion, i) => {
         })
       })
     })
+    it('spawn event', (done) => {
```
Comment (wvffle) [inline test/internalTest.js]: I believe it should be an external test. In #783 someone claimed it happens only on non-vanilla servers. We may add support for non-vanilla tests in the future.

File: lib/plugins/health.js [hunk 3]
```diff
@@ -8,6 +8,12 @@ function inject (bot) {
     bot.emit('respawn')
   })
 
+  bot._client.once('update_health', (packet) => {
+    if (packet.health > 0) {
```
[the remark this answers is not in the document]
Comment (ImHarvol) [reply lib/plugins/health.js]: The bot would [respawn](https://github.com/PrismarineJS/mineflayer/blob/0b9f5010b4c579188827f8977ffc503a2ee3b063/lib/plugins/health.js#L25) and the 'spawn' event would be triggered on https://github.com/PrismarineJS/mineflayer/blob/0b9f5010b4c579188827f8977ffc503a2ee3b063/lib/plugins/health.js#L28

File: test/externalTest.js [hunk 4]
```diff
@@ -58,6 +58,7 @@ mineflayer.supportedVersions.forEach((supportedVersion, i) => {
           version: supportedVersion
         })
         commonTest(bot)
+        bot.test.port = PORT
```
Comment (wvffle) [inline test/externalTest.js]: Why?

File: test/externalTests/spawnEvent.js [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    cb => {
+      // Test spawn event on login
+      const spawnBot = mineflayer.createBot({
+        username: 'spawnbot',
+        viewDistance: 'tiny',
+        port: bot.test.port,
+        host: 'localhost',
+        version: bot.version
+      })
+      spawnBot.once('login', () => {
```
Comment (wvffle) [inline test/externalTests/spawnEvent.js]: This lisstener is not needed

File: test/externalTests/spawnEvent.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        const timeout = setTimeout(() => {
+          spawnBot.removeListener('spawn', spawnHandler)
+          spawnBot.end()
+          assert.fail("spawn event wasn't triggered on connection")
+          cb()
+        }, 2500)
```
Comment (wvffle) [inline test/externalTests/spawnEvent.js]: add a newline after timeout for better looks

File: test/externalTests/spawnEvent.js [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+  const spawnEventTest = [
+    cb => {
+      // Test spawn event on login
+      const spawnBot = mineflayer.createBot({
+        username: 'spawnbot',
+        viewDistance: 'tiny',
+        port: bot.test.port,
+        host: 'localhost',
+        version: bot.version
+      })
```
Comment (wvffle) [inline test/externalTests/spawnEvent.js]: newline

File: test/externalTests/spawnEvent.js [hunk 8]
```diff
@@ -0,0 +1,50 @@
+const assert = require('assert')
+const mineflayer = require('../../')
+
+module.exports = () => (bot, done) => {
+  const spawnHandler = (timeout, client, cb) => {
+    clearTimeout(timeout)
+    if (client) client.end()
+    cb()
+  }
```
Comment (wvffle) [inline test/externalTests/spawnEvent.js]: newline

File: test/externalTests/spawnEvent.js [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+        }, 2500)
+        spawnBot.once('spawn', () => {
+          spawnHandler(timeout, spawnBot, cb)
+        })
+      })
+    },
+    cb => {
+      // Test spawn event on death
+      bot.chat('/kill')
+      bot.once('death', () => {
```
Comment (wvffle) [inline test/externalTests/spawnEvent.js]: This listener is not needed

File: test/externalTests/spawnEvent.js [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+        const timeout = setTimeout(() => {
+          bot.removeListener('spawn', spawnHandler)
+          bot.end()
```
Comment (wvffle) [inline test/externalTests/spawnEvent.js]: I'm not sure if it wont crash the tests

File: test/externalTests/spawnEvent.js [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+          assert.fail("spawn event wasn't triggered on death")
+          cb()
+        }, 2500)
```
Comment (wvffle) [inline test/externalTests/spawnEvent.js]: newline

**Comment by wvffle** [comment]
Resolve all things in the review and check why the tests fail. You can check your tests locally running `mocha test/externalTest.js` if you want to check specific version you can add `-g "1.12"`

**Comment by ImHarvol** [comment]
Thank you! I ran the test locally, and it worked fine. I think the test failed because of the latency, though I'm not sure. I'll take a look tomorrow

**Comment by wvffle** [comment]
LGTM

**Comment by rom1504** [comment]
lgtm

**Pull request merged**